Added lexicographic minimum ratio test to lemke-howson

newaijj opened the pull request:
Read #65 and tried adding it.

Main changes are 
https://github.com/newaijj/Nashpy/blob/ffea3522706ad51f712d42023d41683c8fa740e6/src/nashpy/integer_pivoting/integer_pivoting_lex.py#L1
and
https://github.com/newaijj/Nashpy/blob/ffea3522706ad51f712d42023d41683c8fa740e6/src/nashpy/algorithms/lemke_howson_lex.py#L1

Demo of the additional degenerate games that can be solved are here
https://github.com/newaijj/Nashpy/blob/ffea3522706ad51f712d42023d41683c8fa740e6/tests/unit/test_lemke_howson_lex.py#L9

Not all degenerate games can be solved (example commented out in test script). I'm not sure what kinds of games can be and what kinds can't, so I would really like to find out more about this!

drvinceknight commented:
Thanks for this. 

First thing is that for whatever reason the continuous integration testing hasn't kicked off. I've recently moved to GitHub Actions and that might be why. Could you modify this line https://github.com/drvinceknight/Nashpy/blob/master/.github/workflows/config.yml#L3

To be:

```
on: [push, pull_request]
```

Secondly it looks like your formatter/editor has gone through and changed a number of things. For example: https://github.com/drvinceknight/Nashpy/pull/71/files#diff-52d9477f45c7765aa7185d58871fcd43L8 (changing 4 space tabs to something else). This makes the diffs harder to isolate but also will fail once the continuous integration is working properly as I use `black` to format files.

Finally there are a few files that you've committed that should not be `.DS_Store` in various locations for example. If you could remove them that will be great (don't worry about rebasing the commits I'll squash everything once it's ready to be merged).

Once those things are set and the tests pass I'll take a look more carefully through the code itself but thank you very much for working on it. If anything I've said above is not clear let me know and I'll be happy to help.

newaijj commented:
Yep all tests pass now I guess you can look through the code!

drvinceknight commented:
Great thanks although it looks like you've removed the test file? Also it looks like the diff's are still pretty messy with your reformatting of the style, for example: https://github.com/drvinceknight/Nashpy/pull/71/files#diff-52d9477f45c7765aa7185d58871fcd43R64

If you could fix those things please, if you're unsure how let me know.

newaijj commented:
The test file wasn’t in the same format as the rest of your unit tests, so I thought I would add it another time. I can add it back first then.

newaijj commented:
oookay I removed the tabs/4 spaces diff and I moved the test file to a new demos folder so it doesn't mess up pytest. Sorry about all the messy diffs and repeated commits haha I'm new to the whole CI thing

drvinceknight commented:
Cool thanks for that, we're making progress, first thing lets move the demo file to a test file using the same format as the rest of the tests. Let me know if you need a hand with that.

newaijj commented:
Actually yes can I know if the test will run on my commit? Because I’m not sure how I can check the test i want to add is working properly

drvinceknight commented:
The [README](https://github.com/drvinceknight/Nashpy#development_ has the following information:

## Development

To install a development version of this library:

```
$ python setup.py develop
```

To run the full test suite (install `pytest`: `pip install pytest`) and run:

```
$ pytest .
```

drvinceknight commented:
Thanks for the tests! I've got some time tomorrow to review the contribution in detail. :+1:

newaijj commented:
Alright thanks!

drvinceknight commented on src/nashpy/algorithms/lemke_howson_lex.py:
```diff
@@ -0,0 +1,174 @@
+)
+
+
+def shift_tableau(tableau, shape):
```
I believe this is the same as `lemke_howson.shift_tableau` so let's delete it and import it from there.

drvinceknight commented on src/nashpy/algorithms/lemke_howson_lex.py:
```diff
@@ -0,0 +1,174 @@
+    )
+
+
+def tableau_to_strategy_lex(tableau, basic_labels, strategy_labels):
```
I don't think this differs from `lemke_howson.table_to_strategy` so let's import it from there and delete this one (and rename in the places it's used).

drvinceknight commented on src/nashpy/algorithms/lemke_howson_lex.py:
```diff
@@ -0,0 +1,174 @@
+
+	"""
+
+    if np.min(A) <= 0:
```
There is a lot of duplication in this function, could you modularise it please. You might need to modify `lemke_howson.lemke_howson` to take a pivoting function as an argument.

newaijj replied:
Can I know how you would like lemke_howson to be modularised? Because other than the pivoting function I also added slack variables, changed how the tableaux cycle and how the strategies are calculated. I'm just afraid it might get too messy if I have to change lemke_howson to accommodate these changes

newaijj replied:
And is it ok of my version terminates with an error if the inputs can’t be handled properly? Because yours gives a runtime warning and still outputs things, so i was wondering if that would be better

newaijj replied:
Because of the above issue with the errors I've put off adding the test for a degenerate game from test_lemke_howson to test_lemke_howson_lex as the runtime warning doesn't happen. 

Thanks for all the help so far!

drvinceknight replied:
> Can I know how you would like lemke_howson to be modularised? Because other than the pivoting function I also added slack variables, changed how the tableaux cycle and how the strategies are calculated. I'm just afraid it might get too messy if I have to change lemke_howson to accommodate these changes

I plan on looking through this in more detail on Wednesday. It might be that this gets refactored subsequently.

> And is it ok of my version terminates with an error if the inputs can’t be handled properly? Because yours gives a runtime warning and still outputs things, so i was wondering if that would be better

I'd rather the runtime warning, it's important to stay consistent otherwise this leads to confusion. Changing the overall behaviour could be a conversation for another time.

This will also allow you to add those tests.

As I mentioned above, I've got more time on Wednesday so I'll get to this properly then, in the meantime if you could make the changes I've suggested regarding the runtime warning that would be great. Note also that there is a merge conflict with the main branch, if you could fix that that would be great (amongst other things it'll add a coverage report to this PR). Let me know if I can assist with anything.

drvinceknight replied:
> Note also that there is a merge conflict with the main branch, if you could fix that that would be great (amongst other things it'll add a coverage report to this PR). Let me know if I can assist with anything.

To be clear: I'm asking you to rebase/merge in the current master branch. Let me know if I can assist.

newaijj replied:
Hi sorry but I’ll only be able to access my computer earliest on Friday night, so I won’t be able to make the changes until then.

drvinceknight replied:
No problem, on rush :) :+1:

drvinceknight commented on src/nashpy/algorithms/lemke_howson_lex.py:
```diff
@@ -0,0 +1,174 @@
+        )
+
+    # First pivot (to drop a label)
+    next_tableux = next(tableux)
```
If we keep these lines (see my previous comment): this variable should be `next_tableau` (no `x` as it indicate plurality).

drvinceknight commented on src/nashpy/integer_pivoting/integer_pivoting_lex.py:
```diff
@@ -0,0 +1,108 @@
+import warnings
+
+
+def make_tableau(M):
```
I believe this is a duplicate from `integer_pivoting` let's import from there.

drvinceknight commented on src/nashpy/integer_pivoting/integer_pivoting_lex.py:
```diff
@@ -0,0 +1,108 @@
+
+    Identifies the row to pivot by finding the first fow in a lexicographical 
+    ordering of rows. (First checks minimum ratio test, then uses lexicographical
+     ordering to break ties)
```
Could you add to the docstring please:

- References to text you're basing the implementation on;
- Explanation of variables `C` and `Cq` in particular but possibly more.

newaijj replied:
I added some information here, hopefully it is enough to explain how the pivoting works.

drvinceknight commented on src/nashpy/integer_pivoting/integer_pivoting_lex.py:
```diff
@@ -0,0 +1,108 @@
+    return np.lexsort(np.flipud((filtered_ratio, lex_order)))[0]
+
+
+def non_basic_variables(tableau):
```
I believe this is a duplicate, let's import from the `integer_pivoting`.

drvinceknight commented on tests/unit/test_lemke_howson_lex.py:
```diff
@@ -0,0 +1,51 @@
+
+class TestLemkeHowsonLex(unittest.TestCase):
+
+    # tests with non-degenerate case
```
Could you make this a docstring.

drvinceknight commented on tests/unit/test_lemke_howson_lex.py:
```diff
@@ -0,0 +1,51 @@
+
+    # tests with non-degenerate case
+
+    def test_particular_lemke_howson(self):
```
Rename to:

```
def test_particular_lemke_howson_with_lexicographic_ratio_test(self):
```

drvinceknight commented on tests/unit/test_lemke_howson_lex.py:
```diff
@@ -0,0 +1,51 @@
+                    all(np.isclose(eq, np.array([1 / 2, 1 / 2]))), msg=str(eq)
+                )
+
+    """
```
Move to the docstring of the next function.

drvinceknight commented on tests/unit/test_lemke_howson_lex.py:
```diff
@@ -0,0 +1,51 @@
+    Tests for Lemke Howson with lexographical ordering
+    """
+
+    def test_particular_lemke_howson_lex(self):
```
```
def test_particular_lemke_howson_with_lexicographic_ratio_test_with_degenerate_games(self):
```

newaijj commented:
So I was thinking about why my implementation was running into problems on some degenerate games but not others, and I realised that it wasn't actually because of a problem with the lexicographic minimum ratio test, but rather with how the non-basic variables have been kept track of! I changed this part of the code instead and now it just solves degenerate games properly without any runtime error! I'm not sure if the lexicographic minimum ratio test was even needed for solving degeneracy, but anyway I've changed the test for degenerate cases into a normal test and removed the runtime error parts for degenerate cases since they no longer happen.

I'll make this commit first, then I'll take a look at whether the whole minimum ratio test has any impact on the solving of games. If not, I'll make another commit removing all of the unnecessary parts.

drvinceknight commented:
Thanks for all this @newaijj the progress sounds great.

> and I realised that it wasn't actually because of a problem with the lexicographic minimum ratio test, but rather with how the non-basic variables have been kept track of

Can you clarify one thing: are you saying that it was the way the "non-basic variables are being kept track" of in your new implementation (in `lemke_howson_lex`) or in the original implementation of the Lemke Howson Algorithm?

I'll be looking through this in detail on Wednesday :+1:

newaijj commented:
I did this in lemke_howson_lex, but I think the same can be done in the original lemke_howson

drvinceknight commented:
Ok thanks, I'll take a look properly on Wednesday :+1:

drvinceknight commented on src/nashpy/algorithms/lemke_howson_lex.py:
```diff
@@ -0,0 +1,111 @@
+
+	========
+
+		* Edited such that non-basic variables are kept track of manually and integer pivoting is done using lexicographical ordering
```
Reword the docstring to reflect the lexicographic ordering (don't worry about the non-basic variables).

drvinceknight replied:
I do not fully understand why you've modified things to keep track of the non-basic variables manually. I'm sure you have a good reason could you explain it to me please?

newaijj replied:
I found that in the degenerate case, after pivoting, basic variables can also have columns in the tableau with only one non-zero entry. As a result, the problem with the original method of finding the non-basic variables (i.e. finding the indices of the columns with only one non-zero entry) was erroneously identifying some basic variables as non-basic. Therefore, even though after pivoting there is always one entering variable and one exiting variable, the algorithm sometimes can’t identify them because before and after pivoting they still are ‘non-basic’. 

Tracking the non-basic variables manually remedies this because it will correctly identify the entering variable (the index with only one non-zero entry

drvinceknight replied:
Thanks for the explanation, it's clear now.

drvinceknight commented on src/nashpy/integer_pivoting/integer_pivoting_lex.py:
```diff
@@ -0,0 +1,91 @@
+    tableau, column_index, slack_variables, non_basic_variables
+):
+    """
+    Pivots the tableau and returns the dropped label
```
This docstring might need to be updated if we keep this function as is however if `column_index` is an input to this function why does it need to be an output as well? (it is not being changed.)

newaijj replied:
This was just a convenient way for me to set the old entering_variable as the new just_entered_variable at the same time as i set the new entering_variable, so that the naming of variables in lemke_howson_lex isn’t too confusing. But if you would prefer it to be set in lemke_howson_lex I’ll do it there instead.

drvinceknight replied:
I might again be missing something but I don't think we need this here so I'd prefer it to be set in `lemke_howson_lex`. (Apologies in advance if I see what I'm missing at a later point and ask us to change it back! 🙇 )

drvinceknight commented on src/nashpy/integer_pivoting/integer_pivoting_lex.py:
```diff
@@ -0,0 +1,91 @@
+    """
+
+    basic_variables = set(range(tableau.shape[1] - 1)) - non_basic_variables
+    for i in basic_variables:
```
This will return the last `i` for which this is true but it will nonetheless loop over all of them. Is this the behaviour you want? We could just do:

```
for i in basic_variables[::-1]:
  if tableau[pivot_row_index, i] != 0:
     return  i
```

newaijj replied:
There should only be one i for which the condition is true (because each basic variable has its non-zero element on a different row index), but yes this would be faster

drvinceknight replied:
If there's only one then:

```
for i in basic_variables:
  if tableau[pivot_row_index, i] != 0:
     return  i
```

newaijj replied:
Okie

drvinceknight commented on src/nashpy/integer_pivoting/integer_pivoting_lex.py:
```diff
@@ -0,0 +1,91 @@
+    Cq = tableau[:, -1]
+
+    # catch divide by zero warning
+    with warnings.catch_warnings():
```
Nice :+1:

drvinceknight commented on src/nashpy/algorithms/lemke_howson_lex.py:
```diff
@@ -0,0 +1,111 @@
+    next_tableau = next(tableux)
+
+    entering_label, just_entered_label = pivot_tableau_lex(
+        next_tableau[0], initial_dropped_label, next_tableau[1], next_tableau[2]
```
Instead of using indexing for the elements of `next_tableau` could you explicitly write a line saying:

```
next_tableu, next_slack_variables, next_non_basic_variables = next(tableaux)
```

this would make the code slightly more readable :+1:

newaijj commented:
Hi so I've cleaned up the code and added a few tests for integer_pivoting_lex. Do take a look and lmk if we can commit this!

drvinceknight commented:
> Hi so I've cleaned up the code and added a few tests for integer_pivoting_lex. Do take a look and lmk if we can commit this!

Thanks! I'll take a look on Wednesday :) :+1:

drvinceknight commented:
I'm going to merge this and open a few issues for some refactoring I'd like to do/investigate before a new release.

Thanks for all the work: :+1:

newaijj commented:
Yay! Thanks for all the advice! This was a fun project to work on and i learned a bunch from all the comments you gave.
